contracts/RewardManager.sol

```diff
@@ -0,0 +1,273 @@
+        startDistribution = _updatedStartTime;
+        endDistribution = _updatedEndTime;
+    }
+    
+    function updateUpfrontUnlock(uint256 _newUpfrontUnlock) external 
+    checkPercentages(_newUpfrontUnlock, preMaturePenalty) 
+    onlyOwner
+    {
+        upfrontUnlock = _newUpfrontUnlock;
+    }
+
+    function updateWhitelistAddress(address _excludeAddress, bool status) external onlyOwner{
+        excludedAddresses[_excludeAddress] = status;
+    }
+
+    function updateRewardDistributor(address _distributor, bool status) external onlyOwner{
+        rewardDistributor[_distributor] = status;
+    }
+        
+    function handleRewardsForUser(
+        address user,
+        uint256 rewardAmount,
+        uint256 timestamp,
+        uint256 pid,
+        uint256 rewardDebt
+    ) external {
+        require(rewardDistributor[msg.sender],"Not a valid RewardDistributor");
+        if(rewardAmount > 0){
+            if(excludedAddresses[user]){
+                cnt.safeTransfer(user, rewardAmount);
+            }
+            else{
+                uint256 upfrontAmount = rewardAmount.mul(upfrontUnlock).div(1000);
+                cnt.safeTransfer(user, upfrontAmount);
+                vest(user, rewardAmount.sub(upfrontAmount));
+            }
+        }
+    }
+    
+    function vest(address _user, uint256 _amount) internal {
+        require(_getNow() < startDistribution, "Cannot vest");
+        require(_user != address(0), "Cannot vest for Zero address");
+
+        vestedAmount[_user] = vestedAmount[_user].add(_amount);
+
+        emit Vested(_user, _amount);
+    }
+    
+    /**
+     * @notice Vesting schedule data associated for a user
+     * @dev Must be called directly by the beneficiary assigned the tokens in the schedule
+     * @return totalVested Total vested amount for user
+     * @return totalDrawnAmount total token drawn by user
+     * @return amountBurnt total amount burnt while pre maturely drawing
+     * @return claimable token available to be claimed
+     * @return bonusRewards tokens a user will get if nothing has been withdrawn untill endDistribution
+     * @return stillDue tokens still due (and currently locked) from vesting schedule
+     */
+    function vestingInfo(address _user)
+    public view
+    returns (uint256 totalVested, uint256 totalDrawnAmount, uint256 amountBurnt, uint256 claimable, uint256 bonusRewards, uint256 stillDue) {
+        return (
+        vestedAmount[_user],
+        totalDrawn[_user],
+        burntAmount[_user],
+        _availableDrawDownAmount(_user),
+        bonusReward[_user],
+        _remainingBalance(_user)
+        );
+    }
+
+    function _availableDrawDownAmount(address _user) internal view returns (uint256) {
+        uint256 currentTime = _getNow();
+        if (currentTime < startDistribution) {
+            return 0;
+        } else if (currentTime >= endDistribution || totalDrawn[_user] == vestedAmount[_user]) {
```

Comment:
`totalDrawn[_user]  == vestedAmount[_user]`, then availableDrawDown amount is zero ?

Reply:
updated